### ISSUE TITLE
Multi-file export (from one scad file)

### ISSUE THREAD

**Issue by maxerbubba** [openscad/openscad#1041]
When creating a multi-part model, I find myself adding scaffolding to be able to 'choose' a part and then export them one-at-a-time.  During this process, I also have to make up names as I go along, which is error prone ('gear1.stl', 'gear2.stl').  Perhaps I am using the wrong pattern, but this is what I have been using:
which=0;
if(which==0 || which ==1) cube([10,10,10]);
if(which==0 || which==2) sphere(10);

Could we come up with a new keyword to (a) specify that the child objects would appear in a distinct file when exported, and (b) specify the filename of that file as well?
Example: let's call the new keyword 'part':
part('cubething') cube([10,10,10]);
part('roundthing') sphere(10);
And then, when you File->Export, you would be prompted for a directory, and all the files would be written to the target directory:
C:\target\cubething.stl
C:\target\roundthing.stl

Thoughts?
What to do with unspecified objects? (not under a 'part()' tree, with no filename specified)
Can two trees export to the same object? I  would think so (part('a') cube(); part('a')sphere(); )
Can we optimize F6 render to NOT union separate objects? Probably.

Thanks!
##

**Comment by tcurdt** [comment]
Would also love to see this added.

**Comment by Crazy-Ginger** [comment]
Here a little late but I've written a basic python script to do this. 
https://github.com/BeccaWardle/PyUtilities/blob/main/scad_compiler.py

**Comment by BeccaWardle** [comment]
Here a little late but I've written a basic python script to do this. 
https://github.com/BeccaWardle/PyUtilities/blob/main/scad_compiler.py

**Comment by hadess** [comment]
> Here a little late but I've written a basic python script to do this.
> https://github.com/BeccaWardle/PyUtilities/blob/main/scad_compiler.py

I couldn't figure out how to use the script, from a cursory look, although I'm sure I could reverse engineer the code to figure that out, but more importantly, it doesn't have a license, which means it's not redistributable along with openscad files.

**Comment by t-paul** [comment]
I would avise against using this `-D` behavior, it's defined for variable assignment but currently can be used to pretty much anything else too.

Instead you can use something like, reading from `stdin`:
```
echo "use<n8-shell-with-usb.scad> back_shell();" | openscad -o n8-shell-usb-back.stl -
```
That way the temp script is not even needed.

**Comment by vortex64** [comment]
I would just like to mention the distinction of what I discussed in #5785 "Ungrouped Export", which I believe would be automated mesh separation during export, as opposed to what is discussed here in which the original poster suggest marking or tagging in the code separate objects for later export.

Meshlab, Cura, IdeaMaker and several other programs provide a means of auto ungrouping stl files that have spatially separated parts.

**Comment by nophead** [comment]
NopSCADlib uses a Python script to export multiple STL files separately using the command line. Each part that I want to become an STL is made with a module named with an _stl suffix. The Python framework does a lot of other things, like exporting DXFs and SVGs, making a hierarchical bill of materials, assembly views and a readme file that becomes the assembly instructions for the project, See the example here: https://github.com/nophead/NopSCADlib/tree/master/examples/EnviroPlus

**Comment by gsohler** [comment]
there is an export() function in my fork which can be used to export each
sub part separately in a script

**Comment by vortex64** [comment]
Thanks gsohler.  I am checking out your fork now.

**Comment by gsohler** [comment]
A minimalistic example could be:
* get any installer from http://pythonscad.org/
* use such a python code code

from openscad import *

cu = cube(10)
cy = cylinlinder(4,4)

cu.export("cube.stl") # save an stl file

# save many objects into one 3mf file
export(
{
    "my_cube" : cu,
    "my_cyl": cy
}, "mixed.3mf")

**Comment by nophead** [comment]
@gsohler why do your comments appear in HTML in the emails from GitHub. They look OK here but are near impossible to read in the emails.

![Image](https://github.com/user-attachments/assets/f23b86e5-6f48-4612-a203-66e4322fb823)

**Comment by vortex64** [comment]
I'm not sure that will work as desired gsohler.  

Let us say I make a shape (typically a complex object made from many steps in OpenScad) and in the end I decide I want to arbitrarily chop the item into sections by subtracting a thin plane so I can print the item as separate pieces on small foot print 3d printer.

What is required is to analyze the model and decide what is separate and disconnected from other parts.   

Some discussion along these lines here: https://www.reddit.com/r/algorithms/comments/s5a2ix/separating_objects_in_stl_files_thinking_about_a/ 

One might be able to peek at the code in FreeCAD (or other open source program that can do this) to see how it is performed.   But I suspect it is not trivial.

Certainly doing this for all export types would be nice, but I would be happy if it did this only for the most commonly used .stl case.

**Comment by tcurdt** [comment]
@vortex64 splitting the STL feels overly complicated given that openscad has the parts split already.

**Comment by vortex64** [comment]
tcurdt - OpenScad may have the items split.  But I suspect they are not data structure distinct.  They are just a bunch of triangles is space.

Considering just working on raw .stl files... I think you have to start by sorting out all connected triangles in the file and then write those out as a separate .stl, if there are triangles left in the file then repeat...until no more triangles left.

This, I believe, is what this reddit poster is alluding to.
https://www.reddit.com/r/algorithms/comments/s5a2ix/separating_objects_in_stl_files_thinking_about_a/

**Comment by tcurdt** [comment]
> But I suspect they are not data structure distinct. They are just a bunch of triangles is space.

If all data structures are just triangles in space, does that make graphs nothing more than extremely confused meshes?
Though maybe that's a bit too much philosophy for this ticket ;-)

Either way, splitting STL files does not seem very related to this issue as that's post processing.

**Comment by jordanbrown0** [comment]
Sorry, I didn't "get" the geometric-split vs explicit-split distinction. I'll reopen the other one.

**Comment by vortex64** [comment]
I have has success splitting an .stl with python script mentioned as well as converting OpenSCAD's stl to ascii using python numpy-stl library.

This may be easy to port into OpenSCAD's export function and I will see if I can make progress with this.  Continuing my discussion in #5785

**Comment by gsohler** [comment]
I think i get your issue now.

PythonSCAD is actually missing an algorithm to split a solid into its parts
by triangle inspection and return an array like

parts = my_solid.split_into_parts()

for i, part in parts.enumerate():
    part.export("part%d.stl"%(i))

I could implement your algorithm or implement my O(n) idea.

**Comment by gsohler** [comment]
FYI:

https://imgur.com/a/O2iMSvQ

We have a "separate" function now on board.

**Comment by jordanbrown0** [comment]
Please move the discussion of splitting geometrically-separate objects over to #5785, leaving this one for separating through explicit markup. It was a mistake for me to close #5785.

**Comment by erjoalgo** [comment]
So, is there a way to export multiple parts/modules into separate, named STL files from within openscad?

I understand that it would be possible to refactor each part/module into a separate file, and combine all shared utilities into a shared library file. Or maybe use a make file to output different parts into different files using command line switches. Or a separate python script, etc. But this involves a lot of refactoring and splitting stuff into many different files, which can add unnecessary complexity. Sometimes it just makes sense for all related parts to be defined in the same source file, and many times it is necessary to export them into separate files, e.g. for printing.

**Comment by tcurdt** [comment]
@erjoalgo There are a few options.
What I usually do is to have a script and then call separate modules with different variables defined:

```
  $OPENSCAD \
    --hardwarnings \
    --D BOM=true \
    "$1.scad" \
    --D "$MODULE" \
    --o "$TARGET/$OUT.stl" \
    2>&1 | cat > "$TARGET/$OUT.log"
```

```
MODULE="print()"
```

Would be the `print()` module to be called. Of course you could call different ones.
This isn't pretty, requires re-renders and requires a script - but it works.

**Comment by nophead** [comment]
NopSCADlib exports all the STL files for a project automatically.

**Comment by jordanbrown0** [comment]
There are ways to do it, in a semi-automated way using the command line and a wrapping shell script.  I'm happy to discuss and show how I've done it, but this isn't the right place.  This issue is for hypothetical language extensions to make it easier/better.  Please move "how to" discussions over to the mailing list; see https://openscad.org/community.html#forum ,  For immediate answers you can also try IRC, but that's very dependent on who is paying attention at the moment.

**Comment by keesj** [comment]
I had to solve this issue a while ago and was searching for a solution today... 

The last time I solved it using the children construct( I understood the openscad language contstructs a bit better  back then)

https://github.com/keesj/sas-globe/blob/master/hw/mechanical/rotor/rotator.scad#L142
combined with 
https://github.com/keesj/sas-globe/blob/master/hw/mechanical/rotor/build.sh

allows to export the n'th item in the module

**Comment by tomfelker** [comment]
I've been implementing pretty much exactly this (decade-old) suggestion in a library, although it requires a shell script, so it isn't portable or safe.  Having it actually in the language would be much safer, and would also help for exporting 3MF files with the parts and colors actually explicitly specified, rather than this lazy union thing.

My approach is basically, `export_part(filename="foo.stl") blah();`  and export_part() is a module which:
* by default does nothing
* with one set of variables defined, it just echo()s the filename in a greppable format
* with another set of variables, it emits its children only if the variable matches the filename

Then I have a batch file that exports once just to parse the output to get the file list, and then once for each file.

But yeah - really hacky/ugly to do myself, and it seems I'm not the only one here who's had to hack up stuff like this.

Is there a philosophical opposition to a feature like this, or is it just waiting for someone to do it?

**Comment by t-paul** [comment]
The feature itself is waiting for someone to do it. **BUT**:

It's *NOT* going to be an imperative `export()` function. Scripts will not write files directly as part of the script evaluation.

The way of doing this would be based on a declarative information attached to the design and the export can pick this up and decide what to do. That also included additional meta information depending on the file format, e.g. for 3MF this should be able to switch between a) one file per part and b) all parts in one file, some of the parts may be defined as modifier mesh, e.g. for changing the infill density in that area (at some point it may even define the infill density as hint, but AFAIK there's no common specification for that yet).

Unfortunately this also means, it is not going to be a 2h "I've just added an export function" topic. It probably will need quite some discussion as there have been multiple ideas tossed around in other issues. I'm using exactly the case described here in a manual way, so I'm interested in seeing that happen.

**Comment by erjoalgo** [comment]
If I evaluate a .scad file like this:

    openscad.sh -o output.stl model.scad

 An STL file is written during the evaluation.

Instead of specifying an explicit output file name, the user could specify
a filename prefix or template to accommodate for multiple parts to be
exported to separate files:

    openscad -o output-part-%%% model.scad

Inside model.scad the author would have access to a family of export
functions, eg export_stl, export_3mf, etc. They would all accept a string
literal part name as the first argument, and optional additional arguments
specific to the export file format.

This seems to me like a reasonable proposal that's not incompatible with
the existing design philosophy. It would also be relatively simple to
implement: modify the -o flag to accept a template placeholder, and provide
the export function implementations.

Does this proposal seem reasonable?

Ernesto

**Comment by t-paul** [comment]
> An STL file is written during the evaluation.

No, it is not.

> Inside model.scad the author would have access to a family of export functions, eg export_stl, export_3mf, etc

I specifically said, that is **NOT** going to happen in OpenSCAD scripts themselves.

> This seems to me like a reasonable proposal that's not incompatible with the existing design philosophy.

It is. The scad script is a static, timeless *description* of a 3d object, not an imperative program *generating* a 3d object.

**Comment by erjoalgo** [comment]
In the proposal I made, if we rename "export" to "part", would this now be
compatible with keeping .scad as a static, timeless, non -imperative
description of a 3d object? I'm just declaring that there's a separate part
to the model, not necessarily exporting it. The actual exporting would take
place when the user specifies the -o command line switch, which would be
enhanced to support outputting multiple separate parts instead of just one.

Does this make sense? Or do you have an alternative proposal?

Ernesto

**Comment by t-paul** [comment]
Yes, that would cover the main point. I don't think a potential `part()` implementation can handle the more useful cases, but it at least would handle the naming topic in a way compatible with OpenSCAD script behavior. Note that this also would apply to the GUI export so it's not limited to people using command line.

**Comment by tomfelker** [comment]
I totally agree on having the language side of it be descriptive, rather than imperative.  In some sense it's no different from a color() directive (and I think people have even proposed versions based on abusing that).  Just, instead of "when you render this, make it pink", we'd be saying something like "when you export this, put it in its own 3mf object" or "when exporting this to multiple STL files, use this filename for this object".  `part()` is good for this.  `export()` is perhaps a bit confusing, since it's both a noun and a verb.  Maybe `part_info()` or `export_hint()` would be more clear that it's really serving to add some info into the tree.

One thing that's important to me (and not captured in the script workarounds I've seen, which is why I made my own):  I'd want the set of thingies to be contained only in the .scad file, so that when adding a new thing to be exported, one doesn't have to also add it in some external script, build system, or manual process.

And yeah, certainly nothing should happen while 'evaluating' the file, like echo().  Rather, like color(), it'd add info to the rendered geo, and then either modify existing export processes (like putting multiple objects in one .3MF file), or add some new export processes (e.g., `openscad --export-all --export-dir myexports/`, or erjoalgo's %%% proposal, and corresponding GUI flows).  Whether to include intended output format in the description is an interesting subtlety - I'd lean towards allowing you to hint at it, but also allowing you to override it.

**Comment by t-paul** [comment]
Exactly. As for the more flexible specification of additional data, I had a prototype long ago adding something what other languages call annotations or similar. See https://github.com/openscad/openscad/wiki/Meta-Data-Use-Cases for some examples. The main point of those is that you can attach data, like "infill-density = 0.85" to an object as well.

So I could have (ignore the specific syntax, that's just following the linked wiki article, can be something else easily). Maybe lets keep part() for the naming, it could be that, or some other annotation.

```
part("main")
    cube(20, center = true);

@MeshType(type="modifier")
@SlicerHint(infill_density=0.85)
part("infill-modifier")
    sphere(8);
```

Exporting that to single 3MF could then automatically provide the modifier mesh to the slicer and set the high infill density for the sphere in the center of the cube. The other parts of the cube would just use the default settings from the slicer.

Main point here is that the separate annotations can just be ignored by other exports. So for STL you will simply get 2 files named `main.stl` and `infill-modifier.stl` that you then have to manually load into the slicer and assign the needed properties.

Also the OpenSCAD geometry engine, does not need to know about the annotations (or, maybe just some specific ones), only the export format would consider slicer hints.

**Comment by erjoalgo** [comment]
Maybe the template could specific the file format:

    openscad -o output-part-%%%.stl model.scad

Similar to imagemagick's convert utility:

    convert image.pgn image.jpg

This way the .scad model file would be limited to describing the part and
wouldn't have to be bothered with export or file format specific details.

Ernesto

**Comment by t-paul** [comment]
> Maybe the template could specific the file format:

That is how it works now for the format selection. If for some reason it can't be included in the name, there's also a separate `--export-format` flag already.

**Comment by jordanbrown0** [comment]
Any such scheme needs to describe what happens when the "part" specification is not at the top level...
```
difference() {
     union() {
         part("red") A();
         part("blue") B();
    }
    cube(10);
}
```
and what happens when the parts overlap.

I think that to solve the color/material problem right, there needs to be a real answer, not "it's an error".

A top-level-only, ignore-overlap answer might be a stepping stone, but I don't think it would be a full answer.

**Comment by t-paul** [comment]
Yes, I would start with only allowing this at top-level, ignoring everything else marked deeper in the tree (which is why I'm not liking the `part()` thing as this will prevent any other module to be top-level, at least without other magic like `part()` at top-level means all the modules one level below are still sort-of top-level too).

And we can't force it to be non-overlapping anyway. That would make it impossible to export modifier meshes which are by definition overlapping the "normal" part.

But I fully agree that listing all the "it will be able to do" and also "it won't handle those cases" would be the important first step.

**Comment by jordanbrown0** [comment]
> And we can't force it to be non-overlapping anyway. That would make it impossible to export modifier meshes which are by definition overlapping the "normal" part.

I'm not sure exactly what a "modifier mesh" is.  Could you elaborate a bit on the requirement?

I would hope that eventually
```
union() {
    color("red") cube(10);
    color("blue") sphere(10);
}
```
would produce a model with non-overlapping red and blue components.  (I have a mild preference that the intersection volume be red, but would accept other arguments.)

**Comment by nophead** [comment]
Modifier meshes are used by slicers to apply different settings to part of a model. For example, different infill densities.

**Comment by t-paul** [comment]
Example: https://help.prusa3d.com/article/modifiers_1767#modifier-meshes

**Comment by jordanbrown0** [comment]
That's what I kind of thought, but wanted to be sure.

I'm pretty firm on the notion that multi-color / multi-material models (like the red-cube/blue-sphere model above) should produce disjoint results.

So how do we reconcile that requirement with your requirement?

I come up with two possible approaches.

My first thought was that modifiers would impose sort of a Venn diagram, with the interaction of a "normal" part and a "modifier" part yielding A-B, A-intersect-B, and B-A as three separate results.

But I think my second thought would be better: modifiers live on a different plane of existence, or perhaps multiple different planes of existence, and so simply don't interact in OpenSCAD with "normal" shapes.  Whether they interact in the slicer is up to the slicer.  You could also use this to generate N independent result meshes, for whatever purpose you like.

But again I think that combining a red part with a blue part needs to yield disjoint red and blue results, without requiring the OpenSCAD program to ensure that they do not overlap, and so "red" and "blue" cannot be on different planes of existence.

**Comment by tomfelker** [comment]
It's important to me that this works inside `module()`, so I'd hope any 'top-level-only' restrictions still allowed that.

I'm running into this problem right now, trying to make a multi-color model using lazy-union.  For example, I'd like to do something like this, arranging a print bed with some multi-material nametags on it:

```
module nametag(label)
{
    part("material_1") difference() { cube(); text(label); }
    part("material_2") intersection() { cube(); text(label); }
}

nametag("Alice");
translate([100, 0, 0]) nametag("Bob");
```

but when I do that using lazy-union, I only get one object per top-level module call.  So I'm stuck either duplicating code or adding module arguments everywhere to restructure the whole thing to the point where I can do:

```
nametag("Alice", main_part=true, label_part=false);
nametag("Alice", main_part=false, label_part=true);

translate([100, 0, 0]) nametag("Bob", main_part=true, label_part=false);
translate([100, 0, 0]) nametag("Bob", main_part=false, label_part=true);
```

and then I'm left with complexity that can't be hidden, and where the caller needs to know the full list of parts in advance.

My shell script's approach handles this.  All it's doing is this:
```
module part(part_name)
{
	if(for_export && export_specific_part != "" && export_specific_part == part_name)
	children();

	if(list_exports)
	echo("export_part", part_name);
}
```
and driving those global variables via the shell script to discover the parts and export them each individually.  The limit is that it can't make a single 3MF file, or take advantage of render() caching, since it's invoking OpenSCAD once per part() - but we could fix this by basically bringing the feature into OpenSCAD as an outer loop.

With these semantics, the answer to @jordanbrown0 's question would be that you'd get two parts that are the equivalent of this:

Red part:
```
difference() {
     union() {
         part("red") A();
         if (0) part("blue") B();
    }
    cube(10);
}
```
Blue part:
```
difference() {
     union() {
         if (0) part("red") A();
         part("blue") B();
    }
    cube(10);
}
```

**Comment by t-paul** [comment]
> It's important to me that this works inside `module()`, so I'd hope any 'top-level-only' restrictions still allowed that.

That is meant as a way of getting *something*, not as a permanent restriction. That said, if we just try to find the most complex cases we can't solve, it is sort-of permanent because nothing ever gets implemented.

So that was just a suggestion to go the first step. If we can reduce restrictions on the way early, that's great.

**Comment by jordanbrown0** [comment]
Don't rely on lazy union.  I am not familiar with the details, but people who are familiar with it say that it's dead in the water in its current form.

WRT your wrapper script, I have a very similar one.  I would not consider it to be an acceptable solution to the multicolor problem, because it puts too much onus on the user to keep the parts separate.

**Comment by jordanbrown0** [comment]
> [lazy union] is dead in the water in its current form.

And nobody is working on it.

**Comment by coryrc** [comment]
> I'm running into this problem right now, trying to make a multi-color model using lazy-union.

Have you seen [colorscad](https://github.com/jschobben/colorscad)? I believe it should work with a recent nightly, as I just exported a model to 3mf and could break it up to parts.
